src/anndata/_io/specs/methods.py

```diff
@@ -301,13 +330,22 @@ def write_raw(f, k, raw, _writer, dataset_kwargs=MappingProxyType({})):
 
 @_REGISTRY.register_read(H5Group, IOSpec("dict", "0.1.0"))
 @_REGISTRY.register_read(ZarrGroup, IOSpec("dict", "0.1.0"))
-def read_mapping(elem, _reader):
+def read_mapping(
+    elem: GroupStorageType, *, _reader: Reader
+) -> dict[str, InMemoryArrayOrScalarType]:
     return {k: _reader.read_elem(v) for k, v in elem.items()}
 
 
 @_REGISTRY.register_write(H5Group, dict, IOSpec("dict", "0.1.0"))
 @_REGISTRY.register_write(ZarrGroup, dict, IOSpec("dict", "0.1.0"))
-def write_mapping(f, k, v, _writer, dataset_kwargs=MappingProxyType({})):
+def write_mapping(
+    f: GroupStorageType,
+    k: str,
+    v: dict,
```

Comment:
this should be the same type as above: dict from str to everything we can read/write including itself.